File: 3439_Reschedule_Meetings_for_Maximum_Free_Time_1.js
var maxFreeTime = function(eventTime, k, startTime, endTime) {
    const gaps = [];

    // Gap before first meeting
    gaps.push(startTime[0]);

    // Gaps between meetings
    for (let i = 1; i < startTime.length; i++) {
        gaps.push(startTime[i] - endTime[i - 1]);
    }

    // Gap after last meeting
    gaps.push(eventTime - endTime[endTime.length - 1]);

    // Sliding window of size k+1
    let window = 0;
    for (let i = 0; i <= k; i++) {
        window += gaps[i];
    }

    let maxGap = window;

    for (let i = k + 1; i < gaps.length; i++) {
        window += gaps[i] - gaps[i - (k + 1)];
        maxGap = Math.max(maxGap, window);
    }

    return maxGap;
};

File: 3439_Reschedule_Meetings_for_Maximum_Free_Time_1.java
class Solution {
    public int maxFreeTime(int eventTime, int k, int[] startTime, int[] endTime) {
        List<Integer> gaps = new ArrayList<>();

        // Gap before first meeting
        gaps.add(startTime[0]);

        // Gaps between meetings
        for (int i = 1; i < startTime.length; i++) {
            gaps.add(startTime[i] - endTime[i - 1]);
        }

        // Gap after last meeting
        gaps.add(eventTime - endTime[endTime.length - 1]);

        // Sliding window of size k+1
        int window = 0;
        for (int i = 0; i <= k; i++)
            window += gaps.get(i);

        int maxGap = window;

        for (int i = k + 1; i < gaps.size(); i++) {
            window += gaps.get(i) - gaps.get(i - (k + 1));
            maxGap = Math.max(maxGap, window);
        }

        return maxGap;
    }
}
